src/main/kotlin/three/consulting/epoc/repository/TimesheetEntryRepository.kt

```diff
@@ -37,4 +37,29 @@ interface TimesheetEntryRepository : JpaRepository<TimesheetEntry, Long> {
         @Param("startDate") startDate: LocalDate,
         @Param("endDate") endDate: LocalDate
     ): List<TimesheetEntry>
+
+    @Query(
+        """
+            SELECT te.* from timesheet_entry te
+            LEFT JOIN timesheet t ON t.id = te.timesheet_id
+            LEFT JOIN task tk ON tk.id = te.task_id
+            LEFT JOIN employee e ON e.id = t.employee_id
+            LEFT JOIN project p ON p.id = t.project_id
+            LEFT JOIN customer c ON c.id =  p.customer_id
+            WHERE (te.date >= :startDate AND te.date <= :endDate)
+                AND (:email is null or e.email = :email)
+                AND (:project is null or p.name = :project)
+                AND (:customer is null or c.name = :customer)
+                AND (:task is null or tk.name = :task)
```

Comment:
Bigger issue here is exposing attack surface for sql injection as the string query parameters aren't validated at controller level and can contain any string values, so this is a little dangerous. Why not use id's as those should be just as available to you at the frontend as the names? Then you won't need to figure out proper validations for these name fields and can just use `Long` types.

Reply:
Yeah. This is true. In last project I worked on frontend and backend data was separated and you couldn't use frontend Id:s to find data at backend but there was API for that. Now both data are the same and therefore the Id is the best way to go. Have have to make these changes also to frontend.